src/core/util.ts

```diff
@@ -653,20 +653,70 @@ export function isPrimitive(obj: any): boolean {
     return obj[primitiveKey];
 }
 
+interface MapInterface<T, KEY extends string | number = string | number> {
+    // Size should be a getter here, but tsconfig will not allow getters (es3 target).
+    // Since our polyfill and HashMap class do not expose this, it should be ok.
+    delete(key: KEY): boolean;
+    get(key: KEY): T | undefined
+    set(key: KEY, value: T): this
+    keys(): KEY[];
+    forEach(callback: (key: KEY, value: T) => void): void
+}
+
+
+class MapPolyfill<T, KEY extends string | number = string | number> implements MapInterface<T, KEY> {
+    private data: Record<KEY, T> = {} as Record<KEY, T>;
+
+    delete(key: KEY): boolean {
+        const existed = this.data.hasOwnProperty(key);
+        if (existed) {
+            delete this.data[key];
+        }
+        return existed;
+    }
+    get(key: KEY): T | undefined {
+        return this.data[key];
+    }
+    set(key: KEY, value: T): this {
+        this.data[key] = value;
+        return this;
+    }
+    keys(): KEY[] {
+        return Object.keys(this.data) as KEY[];
+    }
+    forEach(callback: (key: KEY, value: T) => void): void {
+        // This is a potential performance bottleneck, see details in
+        // https://github.com/ecomfe/zrender/issues/965, however it is now
+        // less likely to occur as we default to native maps when possible.
+        for (const key in this.data) {
+            if (this.data.hasOwnProperty(key)) {
+                callback(key, this.data[key]);
+            }
+        }
+    }
+}
+
+// We want to use native Map if it is available, but we do not want to polyfill the global scope
+// in case users ship their own polyfills or patch the native map object in any way.
+const isNativeMapSupported = typeof Map === 'function';
+function maybeNativeMap<T, KEY extends string | number = string | number>(): MapInterface<T, KEY> {
+    // Map may be a native class if we are running in an ES6 compatible environment.
+    // eslint-disable-next-line
+    return (isNativeMapSupported ? Map : MapPolyfill) as unknown as MapInterface<T, KEY>;
+}
 
 /**
  * @constructor
- * @param {Object} obj Only apply `ownProperty`.
+ * @param {Object} obj
  */
 export class HashMap<T, KEY extends string | number = string | number> {
-
-    data: {[key in KEY]: T} = {} as {[key in KEY]: T};
+    data: MapInterface<T, KEY>
 
     constructor(obj?: HashMap<T, KEY> | { [key in KEY]?: T } | KEY[]) {
         const isArr = isArray(obj);
         // Key should not be set on this, otherwise
         // methods get/set/... may be overrided.
-        this.data = {} as {[key in KEY]: T};
+        this.data = maybeNativeMap<T, KEY>();
```

Comment:
The `maybeNativeMap()` seems to return a class instead of an instance.

Reply:
Nice find.